### PR TITLE
:seedling: Add tests for external patch apply

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add tests to Apply using an external Patch.

Part of #6545

### CODE REVIEW

File: internal/controllers/topology/cluster/patches/template.go [hunk 1]
```diff
@@ -57,11 +57,14 @@ func (t *requestItemBuilder) WithHolder(object client.Object, fieldPath string)
 	return t
 }
 
+// uuidGenerator is defined as a package variable to enable changing it during testing.
+var uuidGenerator func() types.UID = uuid.NewUUID
```
Comment (killianmuldoon) [inline internal/controllers/topology/cluster/patches/template.go:61]: This is one of the things I'm not fully comfortable with - it lets us generate the UUID at runtime for testing. It's fully private, but it's not being injected properly. The alternative would be to pass it through a call chain though as this is essentially for matching patches.

Comment (sbueringer) [reply internal/controllers/topology/cluster/patches/template.go:61]: I'm fine with the current implementation and prefer it over alternatives. I've seen this pattern used before and I don't think it's too problematic.

File: internal/controllers/topology/cluster/patches/engine_test.go [hunk 2]
```diff
@@ -49,9 +52,11 @@ func TestApply(t *testing.T) {
 	}
 
 	tests := []struct {
-		name           string
-		patches        []clusterv1.ClusterClassPatch
-		expectedFields expectedFields
+		name                  string
+		patches               []clusterv1.ClusterClassPatch
+		externalPatchResponse map[string]runtimehooksv1.ResponseObject
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/engine_test.go:57]:
```suggestion
		externalPatchResponses map[string]runtimehooksv1.ResponseObject
```
nit

File: internal/controllers/topology/cluster/patches/engine_test.go [hunk 3]
```diff
@@ -412,7 +557,12 @@ func TestApply(t *testing.T) {
 			}
 
 			// Apply patches.
-			g.Expect(patchEngine.Apply(context.Background(), blueprint, desired)).To(Succeed())
+			if err := patchEngine.Apply(context.Background(), blueprint, desired); err != nil {
+				if tt.wantErr != true {
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/engine_test.go:561]:
```suggestion
				if !tt.wantErr {
```
nit

File: internal/controllers/topology/cluster/patches/engine_test.go [hunk 4]
```diff
@@ -348,6 +353,137 @@ func TestApply(t *testing.T) {
...
+						Status:  runtimehooksv1.ResponseStatusFailure,
+						Message: "not a valid infrastructureCluster",
+					},
+				},
+			},
+			expectedFields: expectedFields{
+				infrastructureCluster: map[string]interface{}{
+					"spec.resource": "infraCluster",
+				},
+			},
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/engine_test.go:427]:
```suggestion
```
nit: let's drop the expectedFields as we only expect an error

**Review by sbueringer — COMMENTED** [review]
Thx!

Overall lgtm, a few nits

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**